compiler/AST/astutil.cpp

```diff
@@ -1205,89 +1208,119 @@ visitVisibleFunctions(Vec<FnSymbol*>& fns, Vec<TypeSymbol*>& types)
+  // We delete the reference type if the base type is dead (below).
+  if (ts->hasFlag(FLAG_REF)) return false;
 
-static void pruneUnusedAggregateTypes(Vec<TypeSymbol*>& types)
-{
-  //
-  // delete unused AggregateType types, only deleting references to such
-  // types when the value types are deleted
-  //
-  forv_Vec(TypeSymbol, ts, gTypeSymbols)
-  {
-    // Do not delete reference types.
-    // We delete the reference type if the base type is dead (below).
-    if (ts->hasFlag(FLAG_REF))
-      continue;
+  // Ignore types flagged as primitive.
+  if (ts->hasFlag(FLAG_PRIMITIVE_TYPE)) return false;
 
-    // Ignore types flagged as primitive.
-    if (ts->hasFlag(FLAG_PRIMITIVE_TYPE))
-      continue;
+  // Ignore types that are not class types
+  if (!isAggregateType(ts->type)) return false;
 
-    // Ignore types that are not class types
-    if (!isAggregateType(ts->type))
-      continue;
+  // Do not delete the object class.  It's pretty crucial.
+  if (ts->hasFlag(FLAG_OBJECT_CLASS)) return false;
 
-    // Do not delete the object class.  It's pretty crucial.
-    if (ts->hasFlag(FLAG_OBJECT_CLASS))
-      continue;
+  // Visit only those types not marked as visible.
+  if (types.set_in(ts)) return false;
 
-    // Visit only those types not marked as visible.
-    if (types.set_in(ts))
-      continue;
+  return true;
+}
 
-    ts->defPoint->remove();
+static void pruneUnusedAggregateTypes(Vec<TypeSymbol*>& types) {
+  //
+  // delete unused AggregateType types, only deleting references to such
+  // types when the value types are deleted
+  //
+  forv_Vec(TypeSymbol, ts, gTypeSymbols) {
+    if (shouldPruneAggregateType(ts, types)) ts->defPoint->remove();
   }
 }
 
+static bool shouldPruneRefType(TypeSymbol* ts, Vec<TypeSymbol*>& types) {
+  // This pass, we are interested only in ref types.
+  if (!ts->hasFlag(FLAG_REF)) return false;
 
-static void pruneUnusedRefs(Vec<TypeSymbol*>& types)
-{
-  forv_Vec(TypeSymbol, ts, gTypeSymbols)
-  {
-    // This pass, we are interested only in ref types.
-    if (!ts->hasFlag(FLAG_REF))
-      continue;
+  // Ignore types flagged as primitive.
+  if (ts->hasFlag(FLAG_PRIMITIVE_TYPE)) return false;
 
-    // Ignore types flagged as primitive.
-    if (ts->hasFlag(FLAG_PRIMITIVE_TYPE))
-      continue;
+  // Ignore types that are not class types
+  if (!isAggregateType(ts->type)) {
+    // hilde sez: Ref types are always class types.
+    // So we can't get here.
+    INT_ASSERT(false);
+    return false;
+  }
 
-    // Ignore types that are not class types
-    if (!isAggregateType(ts->type))
-    {
-      // hilde sez: Ref types are always class types.
-      // So we can't get here.
-      INT_ASSERT(false);
-      continue;
-    }
+  // Visit only those types not marked as visible.
+  if (types.set_in(ts)) return false;
 
-    // Visit only those types not marked as visible.
-    if (types.set_in(ts))
-      continue;
+  if (Type* vt = ts->getValType()) {
```

Comment:
If this is null, this function will still return true and attempt to set the ``refType``.

Reply:
Thanks, good catch.